core/lib/via_btc_client/src/inscriber/mod.rs

```diff
@@ -1 +1,734 @@
+
+        let reveal_tx_input_info = self.prepare_reveal_tx_input(
+            &commit_tx_output_info,
+            &final_commit_tx,
+            &inscription_data,
+        )?;
+
+        let reveal_tx_output_info = self
+            .prepare_reveal_tx_output(&reveal_tx_input_info, &inscription_data)
+            .await?;
+
+        let final_reveal_tx = self.sign_reveal_tx(
+            &reveal_tx_input_info,
+            &reveal_tx_output_info,
+            &inscription_data,
+        )?;
+
+        let broadcast_status = self
+            .broadcast_inscription(&final_commit_tx, &final_reveal_tx)
+            .await?;
+
+        if !broadcast_status {
+            return Err(anyhow::anyhow!("Failed to broadcast inscription"));
+        }
+
+        self.insert_inscription_to_context(
+            input,
+            final_commit_tx,
+            final_reveal_tx,
+            commit_tx_output_info,
+            reveal_tx_output_info,
+            commit_tx_input_info,
+        )?;
+
+        Ok(())
+    }
+
+    async fn sync_context_with_blockchain(&mut self) -> Result<()> {
+        if self.context.fifo_queue.is_empty() {
+            return Ok(());
+        }
+
+        let mut new_queue: VecDeque<lib_types::InscriptionRequest> = VecDeque::new();
+
+        let original_queue = self.context.fifo_queue.clone();
+
+        let mut index = 0;
+        while let Some(inscription) = self.context.fifo_queue.pop_front() {
+            let txid_ref = &inscription.fee_payer_ctx.fee_payer_utxo_txid;
+            let res = self
+                .client
+                .check_tx_confirmation(txid_ref, CTX_REQUIRED_CONFIRMATIONS)
+                .await?;
+
+            if index == 0 && !res {
+                // add poped inscription back to the first of queue
+                self.context.fifo_queue = original_queue;
+                break;
+            }
+            if !res {
+                new_queue.push_back(inscription);
+            }
+
+            index += 1;
+        }
+
+        self.context.fifo_queue = new_queue;
+
+        Ok(())
+    }
+
+    async fn prepare_commit_tx_input(&self) -> Result<CommitTxInputRes> {
+        let mut commit_tx_inputs: Vec<TxIn> = Vec::new();
+        let mut unlocked_value: Amount = Amount::ZERO;
+        let mut inputs_count: u32 = 0;
+        let mut utxo_amounts: Vec<Amount> = Vec::new();
+
+        let address_ref = &self.signer.get_p2wpkh_address()?;
+        let mut utxos = self.client.fetch_utxos(address_ref).await?;
+
+        /*
+            adjust utxos list based on unconfirmed utxos in context
+
+            !!! Only Service should send transaction with this address otherwise it will cause a problem in this code !!!
+        */
+
+        let context_queue_len = self.context.fifo_queue.len();
+
+        let mut spent_utxos: HashMap<OutPoint, bool> = HashMap::new();
+
+        for i in 0..context_queue_len {
```

Comment:
let's use iter and enumerate instead.
`for (i, req) in self.context.fifo_queue.iter().enumarete() {`
https://www.oreilly.com/library/view/learning-rust/9781785884306/70af2a22-d0e1-46ab-ae11-72630a834c33.xhtml

Reply:
```
self.context.fifo_queue.get(i).ok_or_else(|| {
                    anyhow::anyhow!("Failed to get inscription request from context")
                })?;
```

i used this to have context error message.